compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/SplitReduction.cpp

```diff
@@ -395,35 +313,91 @@ struct TopkSplitReductionPass
       }
     };
 
-    patterns.add<TopkOpSplitReduction>(
-        patterns.getContext(), splitReductionFn,
-        LinalgTransformationFilter(
-            ArrayRef<StringAttr>{},
-            StringAttr::get(patterns.getContext(), "SPLIT_REDUCTION")));
-    if (failed(applyPatternsAndFoldGreedily(getOperation(),
-                                            std::move(patterns)))) {
-      return signalPassFailure();
-    }
-
-    // Remove all the markers at the end.
+    IRRewriter rewriter(&getContext());
     auto funcOp = getOperation();
-    funcOp->walk([&](TopkOp op) {
-      op->removeAttr(LinalgTransforms::kLinalgTransformMarker);
-      op->removeAttr(kSplitReductionDepthMarker);
-    });
+    SmallVector<LinalgExt::TopkOp> topkCandidates;
+    funcOp->walk([&](LinalgExt::TopkOp op) { topkCandidates.push_back(op); });
+    for (auto op : topkCandidates) {
+      (void)splitReduction(rewriter, op, splitReductionFn);
+    }
   }
 };
 } // namespace
 
-void mlir::iree_compiler::IREE::LinalgExt::populateTopkSplitReductionPattern(
-    RewritePatternSet &patterns,
-    const TopkSplitReductionControlFn &splitReductionFn,
-    const LinalgTransformationFilter &f) {
-  patterns.add<TopkOpSplitReduction>(patterns.getContext(), splitReductionFn,
-                                     f);
+// Transforms an applicable standard single reduction TopkOp into a parallel
+// reduction TopkOp with a reduce step following.
+//
+// Handles parallel reductions in 2 phases: A "map" parallel phase and the a
+// single "reduce" reduction phase. The first phase expands the input tensor
+// shape by breaking the reduction dimension into multiple parallel reductions
+// (upping the rank of the input). Topk is run on these dimensions in parallel
+// The second phase collapses the parallel results into a single final reduce.
+// Topk is run again on the combined output to produce a final output.
+//
+// Currently only topk operations without input indices are supported.
+LogicalResult
+splitReduction(RewriterBase &rewriter, LinalgExt::TopkOp topkOp,
+               const TopkSplitReductionControlFn &splitReductionFn) {
```

Comment:
This part is mostly copied from TopkOpSplitReduction. The difference is

1. It removes all the filter things from the method.
2. It manually does recursive split reduction. In the past, we rely on the driver to apply SplitReduction on new reduction ops. After moving out of filter-based patterns, we need to apply recursive split reduction ourselves.